src/dotnet/CarbonAware.DataSources.WattTime/src/WattTimeDataSource.cs

```diff
@@ -89,30 +91,31 @@ public async Task<EmissionsForecast> GetCurrentCarbonIntensityForecastAsync(Loca
     }
     private async Task<IEnumerable<EmissionsData>> GetCarbonIntensityAsync(Location location, DateTimeOffset periodStartTime, DateTimeOffset periodEndTime)
     {
-        this.Logger.LogInformation("Getting carbon intensity for location {location} for period {periodStartTime} to {periodEndTime}.", location, periodStartTime, periodEndTime);
+        Logger.LogInformation($"Getting carbon intensity for location {location} for period {periodStartTime} to {periodEndTime}.");
 
         using (var activity = ActivitySource.StartActivity())
         {
             BalancingAuthority balancingAuthority = await this.GetBalancingAuthority(location, activity);
-            var data = (await this.WattTimeClient.GetDataAsync(balancingAuthority, periodStartTime, periodEndTime)).ToList();
-
-            Logger.LogDebug("Found {count} total forecasts for location {location} for period {periodStartTime} to {periodEndTime}.", data.Count, location, periodStartTime, periodEndTime);
-
-            // Linq statement to convert WattTime forecast data into EmissionsData for the CarbonAware SDK.
-            var result = data.Select(e => new EmissionsData() 
-            { 
-                Location = e.BalancingAuthorityAbbreviation, 
-                Rating = ConvertMoerToGramsPerKilowattHour(e.Value), 
-                Time = e.PointTime,
-                Duration = FrequencyToTimeSpan(e.Frequency)
-            });
-
+            var data = await this.WattTimeClient.GetDataAsync(balancingAuthority, periodStartTime, periodEndTime);
             if (Logger.IsEnabled(LogLevel.Debug))
             {
-                Logger.LogDebug("Found {count} total emissions data records for location {location} for period {periodStartTime} to {periodEndTime}.", result.ToList().Count, location, periodStartTime, periodEndTime);
+                Logger.LogDebug($"Found {data.Count()} total forecasts for location {location} for period {periodStartTime} to {periodEndTime}.");
+            }
+            if (data.Any())
```

Comment:
This approach definitely solves for the 0 data points case, but I was expecting us to also resolve the case where we ensure we capture all relevant data points, which I don't believe this will do.

for 5 minute interval data, a request for 00:00:01 - 00:05:00 will return 1 datapoint representing the period 00:05:00 - 00:9:59 and the code as written will simply return that 1 point it meets our minimum sampling window.

Is there a reason we dont simply request the minimum sampling window upfront and then filter down from there?

Reply:
> This approach definitely solves for the 0 data points case, but I was expecting us to also resolve the case where we ensure we capture all relevant data points, which I don't believe this will do.
> 
> for 5 minute interval data, a request for 00:00:01 - 00:05:00 will return 1 datapoint representing the period 00:05:00 - 00:9:59 and the code as written will simply return that 1 point it meets our minimum sampling window.
> 
> Is there a reason we dont simply request the minimum sampling window upfront and then filter down from there?

@bderusha , not sure I understand the sentence 'all relevant data points'? Both data sources returns a set of data on a window of time, there is no more data since it is discrete data(not continuous data). Do you want us to 'repeat' data so it would be 'continuous'? No need in my opinion since the time window (it is not in this PR) can be calculated after to improve other granular data (i.e average).

Reply:
Dy all relevant data points I mean that a user requesting data about 00:00:01 through 00:05:00 really wants the point

```
Time = 00:00:00
Duration = 300s
Rating = 123.45
```

but right now and as written here we will still only return what WattTime gave us back, which is:

```
Time = 00:05:00
Duration = 300s
Rating = 131.31
```

If the goal is to ensure that we have datapoints covering the request, I guess I just expected us to grab a larger dataset first and then check for those points.  Doesn't this code effectively do the same thing?  

```
var expandedStart = periodStartTime - TimeSpan.FromMinutes(MinSamplingWindow);
var data = await this.WattTimeClient.GetDataAsync(balancingAuthority, expandedStart, periodEndTime);
return ConvertToEmissionData(data).Select(d => (d.Time + d.Duration) >= periodStartTime && d.Time <=periodEndTime);
```

I think I was surprised to see so much code in this PR based on my understanding of the problem.  Can you help me understand what I am missing?

Reply:
Short answer: Json data doesn't have frequency embedded. Also both source brings data differently, so these amount of code is to centralized that need regardless of the source.

Reply:
Updated version up that I think resolves this issue, added a default frequency value